src/outcomes.py

```diff
@@ -1,39 +1,127 @@
+# pandas dataframes and return a Result object
+def outcome_wrapper(func):
+    def wrapper(preferences: UnionArray, *args, **kwargs):
+        if isinstance(preferences, pd.DataFrame):
+            s = preferences.to_numpy()
+        else:
+            s = preferences
+
+        res = func(s, *args, **kwargs)
+
+        if isinstance(res, dict):
+            all_alternatives = np.unique(preferences)
+            result_dict = {key: res.get(key, 0) for key in all_alternatives}
+            return Result(result_dict)
+
+        return res
+
+    return wrapper
+
+
+# Voting for one
+@outcome_wrapper
+def plurality_outcome(preferences: UnionArray) -> Result:
+    first_prefs = preferences[0, :]
     alternatives, votes = np.unique(first_prefs, return_counts=True)
-    winners_i = np.argwhere(votes == np.max(votes)).flatten()
-    winners = [(alternatives[i], np.max(votes)) for i in winners_i]
-    winners = tie_votes_resolver(winners)
-    return winners
+    results = dict(zip(alternatives, votes))
 
+    return results
 
-def for_two_outcome(s: np.array) -> List[Tuple[str, int]]:
-    return [()]
 
+# Voting for two
+@outcome_wrapper
+def for_two_outcome(preferences: UnionArray) -> Result:
+    first_prefs = preferences[0, :]
+    second_prefs = preferences[1, :]
 
-def veto_outcome(s: np.array) -> str:
-    return ""
+    first_votes = dict(zip(*np.unique(first_prefs, return_counts=True)))
+    second_votes = dict(zip(*np.unique(second_prefs, return_counts=True)))
 
+    results = {
+        k: first_votes.get(k, 0) + second_votes.get(k, 0)
+        for k in first_votes.keys() | second_votes.keys()
+    }
 
-def borda_outcome(s: np.array) -> str:
+    return results
 
-    return ""
 
+# Voting for veto
+@outcome_wrapper
+def veto_outcome(preferences: UnionArray) -> Result:
+    last_prefs = preferences[:-1, :]
+    alternatives, votes = np.unique(last_prefs, return_counts=True)
+    results = dict(zip(alternatives, votes))
 
-def tie_votes_resolver(r: np.array):
-    return r
+    return results
 
 
-def all_schemas_outcomes(s: np.array) -> dict:
+# Borda voting
+@outcome_wrapper
+def borda_outcome(preferences: UnionArray) -> Result:
+    n, m = preferences.shape
+    alternatives = np.unique(preferences)
+    borda_points = {a: 0 for a in alternatives}
+
+    for i in range(n):
+        row = preferences[i, :]
+        alternatives, votes = np.unique(row, return_counts=True)
+        for a in alternatives:
+            borda_points[a] += (n - 1 - i) * votes[np.where(alternatives == a)][0]
+
+    return borda_points
+
+
+# Getting the outcomes for all the voting schemas
+def all_schemas_outcomes(s: UnionArray) -> Result:
     scheme_func = {
         "Plurality Voting": plurality_outcome,
         "Voting for Two": for_two_outcome,
         "Veto Voting": veto_outcome,
         "Borda Voting": borda_outcome,
     }
+
     scheme_outcome = {n: f(s) for n, f in scheme_func.items()}
     return scheme_outcome
+
+
+if __name__ == "__main__":
```

Comment:
should this not be in main.py file

Reply:
No, this code gets executed only when the outcomes.py file is executed directly. It is not directly part of the project and is used to test the function inside outcomes.py

Reply:
ah ok got it thanks